certbot/certbot
# Certbot for Windows auto-update mechanism

This PR implements a mechanism to make Certbot for Windows update itself during the scheduled renew task run twice a day. 

Technically this PR blocks on the signature of the installer executables, but all of this will be added soon, so we can start the review. 

The mechanism is written into pure Powershell. I did that for the following reasons:
* other programs in the wild usually use an approach of a dedicated executable outside the program itself
* because Windows is Windows, it is just impossible for a program to self update: any opened file descriptor could block the process. So at least a part of the logic needs to be outside Certbot. Some can be in it (like checking new version availability, downloading the installer), but it was a minor part, and having the logic splitted in two different languages would decrease the maintainability in my opinion
* Powershell includes all high level features required to make the script short and readable
* `exec` Linux style is not possible on Windows, so a process (like Certbot) cannot transform itself into another process (like the upgrade mechanism), so a wrapper on top of Certbot would be required. With the limitations of Windows I do not think it would make something very elegant. 
* this way the mechanism can be plugged freely anywhere. I think it is relevant to associate it to the scheduled renew task, to avoid to break it if the upgrade would happen in parallel to `certbot renew` call. Here it is not possible, because the renew and the upgrade are called sequentially one after the other. 

As said, the upgrade script is plugged into a specific action in the existing scheduled task. Actions in a task are run sequentially by construction. Also, it is not the script itself that is run, but a copy of it outside Certbot install directory, here again to avoid errors when trying to delete an open file during the upgrade. 

The script itself will:
* get current version of Certbot installed locally
* get remote latest version using the Github API
* download if necessary the installer of this new version
* check it is signed correctly
* check it is signed by us
* run the installer in silent mode, using the original installation directory specified by the user when Certbot has been installed the first time

Potential improvements : check that Certbot is effectively installed with the new version, fallback to previous version otherwise.
